cluster/store/bootstrap.go

```diff
@@ -65,16 +65,38 @@ func (b *Bootstrapper) Do(ctx context.Context, serverPortMap map[string]int, lg
 		case <-ctx.Done():
 			return ctx.Err()
 		case <-ticker.C:
+			// If we have found no other server, there is nobody to contact
+			if len(servers) <= 0 {
+				continue
+			}
 			// try to join an existing cluster
-			if leader, err := b.join(ctx, servers, voter); err == nil {
-				lg.WithField("leader", leader).Info("successfully joined cluster")
+			if leader, err := b.join(ctx, servers, voter); err != nil {
+				lg.WithFields(logrus.Fields{
+					"servers": servers,
+					"action":  "bootstrap",
+					"voter":   voter,
+				}).WithError(err).Warning("failed to join cluster, will notify next if voter")
+			} else {
+				lg.WithFields(logrus.Fields{
+					"action": "bootstrap",
+					"leader": leader,
+				}).Info("successfully joined cluster")
 				return nil
 			}
 
 			if voter {
 				// notify other servers about readiness of this node to be joined
 				if err := b.notify(ctx, servers); err != nil {
-					lg.WithField("servers", servers).WithError(err).Error("notify all peers")
+					lg.WithFields(logrus.Fields{
+						"action":  "bootstrap",
+						"servers": servers,
+					}).WithError(err).Error("notify all peers")
+				} else {
+					lg.WithFields(logrus.Fields{
+						"action":  "bootstrap",
+						"servers": servers,
+					}).Info("notified peers this node is ready to join as voter")
+					return nil
```

Comment:
**issue (blocking):** that will break the loop, i think we break in join because bootstrap is done